### ISSUE TITLE
test

### ISSUE THREAD

**Issue by kiyaaabdian** [RuiZhang97/ISNet#14]
hi
I train your code but I have an issue. how can I test the train model? you don't write demo or test code for it.

**Comment by IDWTLY** [comment]
import torch
import numpy as np
import cv2
import os
from torch.utils.data import Dataset, DataLoader
from torchvision import transforms
from sklearn.metrics import jaccard_score
from tqdm import tqdm
import warnings
warnings.filterwarnings("ignore")

# ===================== 配置参数 =====================
class TestConfig:
    data_root = "./dataset/test"      # 测试集路径
    img_dir = "images"               # 图像文件夹
    mask_dir = "masks"               # 掩码文件夹
    output_dir = "./results_test"         # 预测结果输出目录
    img_size = (512, 512)            # 输入尺寸（根据需求调整）
    batch_size = 4                   # 批处理大小
    device = "cuda" if torch.cuda.is_available() else "cpu"
    model_path = "your pkl"   # 模型路径（改用权重文件）

# ===================== 数据加载器 =====================
class ISNetTestDataset(Dataset):
    def __init__(self, cfg):
        self.img_paths = sorted([
            os.path.join(cfg.data_root, cfg.img_dir, f) 
            for f in os.listdir(os.path.join(cfg.data_root, cfg.img_dir))
            if f.lower().endswith(('.png', '.jpg', '.jpeg'))
        ])
        self.mask_paths = sorted([
            os.path.join(cfg.data_root, cfg.mask_dir, f) 
            for f in os.listdir(os.path.join(cfg.data_root, cfg.mask_dir))
            if f.lower().endswith(('.png', '.jpg', '.jpeg'))
        ])
        self.transform = transforms.Compose([
            transforms.ToTensor(),
            transforms.Resize(cfg.img_size),
            transforms.Normalize(mean=[0.485, 0.456, 0.406], 
                                 std=[0.229, 0.224, 0.225])
        ])
        
    def __len__(self):
        return len(self.img_paths)
    
    def _compute_gradient(self, img):
        """ 生成3通道梯度图（修复通道不匹配问题）[1](@ref) """
        gray = cv2.cvtColor(img, cv2.COLOR_RGB2GRAY)
        grad_x = cv2.Sobel(gray, cv2.CV_64F, 1, 0, ksize=3)
        grad_y = cv2.Sobel(gray, cv2.CV_64F, 0, 1, ksize=3)
        grad_mag = np.sqrt(grad_x**2 + grad_y**2)
        grad_norm = (grad_mag / grad_mag.max() * 255).astype(np.uint8)
        # 复制为3通道（适配模型输入）[1](@ref)
        return np.repeat(grad_norm[:, :, np.newaxis], 3, axis=2)

    def __getitem__(self, idx):
        img = cv2.cvtColor(cv2.imread(self.img_paths[idx]), cv2.COLOR_BGR2RGB)
        mask = cv2.imread(self.mask_paths[idx], cv2.IMREAD_GRAYSCALE) / 255.0
        grad_img = self._compute_gradient(img)  # 生成3通道梯度图
        
        img_tensor = self.transform(img)
        grad_tensor = transforms.ToTensor()(grad_img)
        grad_tensor = transforms.Resize(self.transform.transforms[1].size)(grad_tensor)
        
        return {
            "image": img_tensor,
            "grad": grad_tensor,
            "mask": torch.tensor(mask, dtype=torch.float32),
            "name": os.path.basename(self.img_paths[idx])
        }

# ===================== 评估指标计算 =====================
def compute_metrics(pred, target):
    pred_bin = (pred > 0.5).astype(np.uint8)
    target_bin = target.astype(np.uint8)
    tp = np.sum(pred_bin * target_bin)
    fn = np.sum(target_bin) - tp
    fp = np.sum(pred_bin) - tp
    iou = jaccard_score(target_bin.flatten(), pred_bin.flatten(), zero_division=0)
    pd = tp / (tp + fn + 1e-6)  # 检测率
    fa = (fp / np.prod(pred_bin.shape)) * 1e6  # 每百万像素误报数
    return iou, pd, fa

# ===================== 主测试函数 =====================
def run_test():
    cfg = TestConfig()
    os.makedirs(cfg.output_dir, exist_ok=True)
    
    # 1. 加载模型（修复weights_only兼容性问题）[2](@ref)
    try:
        # 方案1：直接加载权重（推荐）
        from models.isnet import ISNet  # 导入模型定义
        model = ISNet()
        state_dict = torch.load(cfg.model_path, map_location=cfg.device)
        model.load_state_dict(state_dict)
        model.to(cfg.device).eval()
        print(f"✅ 模型加载成功 | Device: {cfg.device}")
    except Exception as e:
        # 方案2：完整模型加载（旧版兼容）
        try:
            model = torch.load(cfg.model_path, map_location=cfg.device)  # 移除weights_only
            model.to(cfg.device).eval()
            print("⚠️ 使用旧版加载方式 | 建议转换为state_dict格式")
        except:
            print(f"❌ 模型加载失败: {str(e)}")
            return

    # 2. 准备数据
    try:
        dataset = ISNetTestDataset(cfg)
        loader = DataLoader(dataset, batch_size=cfg.batch_size, shuffle=False)
        print(f"✅ 数据加载成功 | 样本数: {len(dataset)}")
    except FileNotFoundError:
        print(f"❌ 数据路径错误: {cfg.data_root}")
        return

    # 3. 测试循环
    results = []
    total_metrics = {"iou": 0, "pd": 0, "fa": 0}
    
    with torch.no_grad():
        for batch in tqdm(loader, desc="🚀 推理进度"):
            images = batch["image"].to(cfg.device)
            grads = batch["grad"].to(cfg.device)
            true_masks = batch["mask"].numpy()
            
            # 前向传播（处理元组输出）[3](@ref)
            preds = model(images, grads)
            if isinstance(preds, tuple):  # 模型返回多个输出
                preds = preds[0]          # 取主输出
            preds = torch.sigmoid(preds).cpu().numpy()
            
            for i in range(len(batch["name"])):
                pred_mask = preds[i].squeeze()
                true_mask = true_masks[i].squeeze()
                iou, pd, fa = compute_metrics(pred_mask, true_mask)
                
                # 记录结果
                results.append({
                    "name": batch["name"][i],
                    "iou": iou,
                    "pd": pd,
                    "fa": fa
                })
                total_metrics["iou"] += iou
                total_metrics["pd"] += pd
                total_metrics["fa"] += fa
                
                # 保存预测掩码
                cv2.imwrite(
                    os.path.join(cfg.output_dir, f"pred_{batch['name'][i]}"), 
                    (pred_mask > 0.5).astype(np.uint8) * 255
                )
    
    # 4. 结果汇总
    num_samples = len(results)
    avg_iou = total_metrics["iou"] / num_samples
    avg_pd = total_metrics["pd"] / num_samples
    avg_fa = total_metrics["fa"] / num_samples
    
    print("\n" + "="*55)
    print(f"📊 测试完成 | 样本数: {num_samples}")
    print(f"▪ 平均IoU: {avg_iou:.4f} | ▪ 检测率(Pd): {avg_pd:.4f} | ▪ 虚警率(Fa): {avg_fa:.1f} px/Mp")
    print("="*55)
    
    # 保存详细结果
    with open(os.path.join(cfg.output_dir, "metrics.csv"), "w") as f:
        f.write("filename,iou,pd,fa\n")
        for r in results:
            f.write(f"{r['name']},{r['iou']:.4f},{r['pd']:.4f},{r['fa']:.1f}\n")
    
    return results

# ===================== 执行入口 =====================
if __name__ == "__main__":
    print("="*55)
    print(f"🔥 ISNet模型测试启动 | 输入尺寸: {TestConfig.img_size}")
    print("="*55)
    test_results = run_test()


This may be useful for your testing
